Yosemite/Yosemite/Stores/CardPresentPaymentStore.swift

```diff
@@ -540,6 +563,45 @@ private extension CardPresentPaymentStore {
         storage.deleteObject(storageAccount)
         storage.saveIfNeeded()
     }
+
+    func upsertChargeInBackground(readonlyCharge: WCPayCharge) {
+        let storage = storageManager.viewStorage
+        let storageWCPayCharge = storage.loadWCPayCharge(siteID: readonlyCharge.siteID, chargeID: readonlyCharge.id) ??
+        storage.insertNewObject(ofType: Storage.WCPayCharge.self)
+
+        switch readonlyCharge.paymentMethodDetails {
+        case .cardPresent(let details):
+            let storageCardPresentDetails = storageWCPayCharge.cardPresentDetails ??
+            storage.insertNewObject(ofType: Storage.WCPayCardPresentPaymentDetails.self)
+            let storageReceiptDetails = storageCardPresentDetails.receipt ?? storage.insertNewObject(ofType: Storage.WCPayCardPresentReceiptDetails.self)
+            storageCardPresentDetails.update(with: details)
+            storageReceiptDetails.update(with: details.receipt)
+            storageCardPresentDetails.receipt = storageReceiptDetails
+            storageWCPayCharge.cardPresentDetails = storageCardPresentDetails
+            storageWCPayCharge.cardDetails = nil
+        case .card(let details):
+            let storageCardDetails = storageWCPayCharge.cardDetails ?? storage.insertNewObject(ofType: Storage.WCPayCardPaymentDetails.self)
+            storageCardDetails.update(with: details)
+            storageWCPayCharge.cardDetails = storageCardDetails
+            storageWCPayCharge.cardPresentDetails = nil
+        case .unknown:
+            storageWCPayCharge.cardDetails = nil
+            storageWCPayCharge.cardPresentDetails = nil
+            break
+        }
+
+        storageWCPayCharge.update(with: readonlyCharge)
```

Comment:
Shouldn't this be calling `saveIfNeeded()` too?

Reply:
I'm going to dig in to this a little more as some research and generally acquainting myself better with our core data stack. It looks about 50/50 in existing `upsert` methods on whether we save, either with `saveIfNeeded` or `saveDerivedType`. Adding this meant that it never displayed card details on #6130, but #6164 works consistently still.